rust/lance/src/dataset/write/commit.rs

```diff
@@ -542,6 +556,117 @@ mod tests {
+            .with_params(&write_params)
+            .execute(vec![data])
+            .await
+            .unwrap();
+
+        io_tracker.incremental_stats(); // Reset the stats
+        let read_version = dataset.manifest().version;
+        let _ = CommitBuilder::new(Arc::new(dataset))
+            .execute(sample_transaction(read_version))
+            .await
+            .unwrap();
+
+        // Assert io requetss
+        let io_stats = io_tracker.incremental_stats();
+        assert_eq!(io_stats.read_iops, 0);
+        assert_eq!(io_stats.write_iops, 2);
+        // We can't write them in parallel. The transaction file must exist before
+        // we can write the manifest.
+        assert_eq!(io_stats.num_hops, 2);
+    }
+
+    #[tokio::test]
+    #[rstest::rstest]
+    async fn test_commit_conflict_iops(#[values(true, false)] use_cache: bool) {
+        let cache_size = if use_cache { 10_000 } else { 0 };
+        let session = Arc::new(Session::new(0, cache_size, Default::default()));
+        let io_tracker = Arc::new(StatsHolder::default());
+        let write_params = WriteParams {
+            store_params: Some(ObjectStoreParams {
+                object_store_wrapper: Some(io_tracker.clone()),
+                ..Default::default()
+            }),
+            session: Some(session.clone()),
+            ..Default::default()
+        };
+        let data = RecordBatch::try_new(
+            Arc::new(ArrowSchema::new(vec![ArrowField::new(
+                "a",
+                DataType::Int32,
+                false,
+            )])),
+            vec![Arc::new(Int32Array::from(vec![0; 5]))],
+        )
+        .unwrap();
+        let mut dataset = InsertBuilder::new("memory://")
+            .with_params(&write_params)
+            .execute(vec![data])
+            .await
+            .unwrap();
+        let original_dataset = Arc::new(dataset.clone());
+
+        // Create 3 other transactions that happen concurrently.
+        let num_other_txns = 3;
+        for _ in 0..num_other_txns {
+            dataset = CommitBuilder::new(original_dataset.clone())
+                .execute(sample_transaction(dataset.manifest().version))
+                .await
+                .unwrap();
+        }
+
+        let _ = CommitBuilder::new(original_dataset.clone())
+            .execute(sample_transaction(original_dataset.manifest().version))
+            .await
+            .unwrap();
+
+        let io_stats = io_tracker.incremental_stats();
+
+        // If there is a conflict with two transaction, the retry should require io requests:
+        // * 1 list version
+        // * num_other_txns read manifests (cache-able)
+        // * num_other_txns read txn files (cache-able)
+        // * 1 write txn file
+        // * 1 write manifest
+        // For total of 3 + 2 * num_other_txns io requests. If we have caching enabled, we can skip 2 * num_other_txns
+        // of those. We should be able to read in 5 hops.
```

Comment:
This might seem like a bit of a tall order, but I think it's possible.

For the uncached case, we'd need to optimize this part:
https://github.com/lancedb/lance/blob/c39b7e7a271eb81078e0a404361a54082289b94e/rust/lance/src/io/commit.rs#L96-L108

Ideally we can re-use the manifest size found in the list versions, and then read the manifest in 1 request.

Reply:
I was able to achieve the goal at the cached case, but for uncached, looks like we will repeatedly make some calls for listng and reading, I have not fixed that yet, will create an issue to track it.